### PR TITLE
treewide: clean up architecture specific toolchain setup

### PR DESCRIPTION
**Pull request by maribu**
### Contribution description

Architecture specific toolchain setup is done mostly in `makefiles/arch/<ARCH>.inc.mk`, except for ESP and ARM9. Those oddballs are changed to the common pattern for consistency.

Secondly, each CPU family included the architecture specific settings on its. However, as the architecture was is already known once the features are resolved, one can simple include `$(RIOTMAKE)/arch/$(CPU_ARCH).inc.mk` directly in `$(RIOTBASE)/Makefile.include`.

### Testing procedure

If the toolchains are not properly set up, Murdock will complain. In addition, it might make sense to pick one example application and board for each platform and confirm that binaries doesn't change.

### Issues/PRs references

One step towards making https://github.com/RIOT-OS/RIOT/pull/15993 possible without reordering the processing of `Makefile.dep` and `Makefile.include`.

### CODE REVIEW

File: cpu/arm7_common/Makefile.features [hunk 1]
```diff
@@ -1,5 +1,5 @@
-CPU_ARCH = armv4t
-CPU_CORE = arm7tdmi_s
+CPU_ARCH := armv4t
+CPU_CORE := arm7tdmi_s
```
Comment (fjmolinas) [inline cpu/arm7_common/Makefile.features:2]: Why do we need immediate expansion here?

Comment (maribu) [reply cpu/arm7_common/Makefile.features:2]: You should ask the other way round: Why do we need recursive expansion? We decided to use immediate expansion, unless there are good reasons to differently.

Comment (fjmolinas) [reply cpu/arm7_common/Makefile.features:2]: > We decided to use immediate expansion

Who is we? https://doc.riot-os.org/build-system-basics.html#board-cpu-features, https://doc.riot-os.org/build-system-basics.html#board-cpu-features, https://github.com/RIOT-OS/RIOT/pull/11099#discussion_r289820421, https://github.com/RIOT-OS/RIOT/issues/8913, https://github.com/RIOT-OS/RIOT/issues/10850. AFAIK, past efforts were header the other way.

Comment (benpicco) [reply cpu/arm7_common/Makefile.features:2]: The argument raised in https://github.com/RIOT-OS/RIOT/pull/11099#discussion_r289820421 only applies for non-const variables though.
It makes sense to use `:=` for constants and only for constants.

Comment (fjmolinas) [reply cpu/arm7_common/Makefile.features:2]: Here it does not matter BTW, but seems we have contradicting impressions on the default behavior to adopt.

Comment (fjmolinas) [reply cpu/arm7_common/Makefile.features:2]: > The argument raised in [#11099 (comment)](https://github.com/RIOT-OS/RIOT/pull/11099#discussion_r289820421) only applies to non-const variables.
> It makes sense to use `:=` for constants and only for constants.

Not necessarily, you might still want to override a constant, if you do `CPU_MODEL := foo`, them `CPU_MODULE=bar` won't override it. Again here it probably does not matter, no one is going to do `CPU_ARCH=far make`., and `CPU_MODEL=foo` has the same effect in this context, and both can be overridden by make arguments.

I also meant to reference this comment https://github.com/RIOT-OS/RIOT/pull/11099#discussion_r289842650.

Comment (fjmolinas) [reply cpu/arm7_common/Makefile.features:2]: I'm not trying to bikeshed, the PR looks fine, just seems we are rowing in opposite directions, and I would rather we align :)

Comment (maribu) [reply cpu/arm7_common/Makefile.features:2]: > := has some disadvantages:
>
> * The value of variables defined with := depends on the order they are defined.

That clearly is a pro, not a con. A simple example:

```Makefile
a = foo
b = $(a)
a += bar
```

It is absolutely contrary to the behavior in every sane programming language, that `b` now has the value `foo bar`, and not `foo`.

> Here it does not matter BTW, but seems we have contradicting impressions on the default behavior to adopt.

We had this discussion a few times now. The first time was when the `make info-boards-supported` was taking minutes to completed with CPU usage at 100%, because of the explosive growth of complexity when expressions start to depend on each other. Switching to immediate expansion restored not only normal build time, but also forced some structuring that resulted in much clearer `Makefile` (or, less horrible, to be honest).

I was under the impression that each time we agreed on using immediate expansion by default.

Comment (maribu) [reply cpu/arm7_common/Makefile.features:2]: > I also meant to reference this comment [#11099 (comment)](https://github.com/RIOT-OS/RIOT/pull/11099#discussion_r289842650).

Honestly, recursive evaluation played a big role in why some parts of the build system are so hard to follow and horrible to debug: For understanding what a specific expression does, you do not only need to know what the contents of every referenced variable currently consist of, you also have to track what they *will* look like *every time* the variable is referenced in the future. With the complexity of RIOT's build system, this is impossible to follow. And this is why people start screaming when somebody touches the build system, as everybody knows this is likely to break things at places that seem totally unrelated for reasons nobody could have even dreamed of before.

Using immediate evaluations instead makes things predictable. And I think we all agree that a more predictable build system would greatly benefit us all.

And the argument regarding the better performance of recursive expansion: The exact opposite is true.

Comment (fjmolinas) [reply cpu/arm7_common/Makefile.features:2]: > That clearly is a pro, not a con. A simple example:

It depends, on a recursive build system when the order of inclusion is not clear. Because you will have different behavior in different scenarios.

> much clearer Makefile (or, less horrible, to be honest).

I completely agree here, having it not be horrible would require someone constantly working on build-system stuff IMO.

>We had this discussion a few times now. The first time was when the make info-boards-supported was taking minutes to complete with CPU usage at 100%, because of the explosive growth of complexity when expressions start to depend on each other.

I'm in no case arguing against the use of `:=` in that situation. But the thing is `:=` can also have an overhead, and the question is what is the best default, not for someone that has suffered from the build system enough to decide and read make documentation, but for programmer/contributor lambda, that is focused on getting his c program to compile.

> I was under the impression that each time we agreed on using immediate expansion by default.

I think I'm still formatted to the default of not liking when I see a `:=`, also because when I started looking at build system stuff order of inclusion was still a huge issue, maybe that is no longer the bigger issue and I need to be re-formatted hahaha.

Anyway,  seems both of you think `:=` is the best default. So I'll shut up.

File: cpu/esp_common/Kconfig [hunk 2]
```diff
@@ -9,6 +9,14 @@ config CPU_ARCH_XTENSA
     select HAS_ARCH_32BIT
     select HAS_ARCH_ESP
 
+config CPU_ARCH_XTENSA_LX106
```
Comment (fjmolinas) [inline cpu/esp_common/Kconfig:12]: I think these changes should be In another commit.

File: cpu/arm7_common/Makefile.features (hunk 1 above)
Comment (fjmolinas) [inline cpu/arm7_common/Makefile.features:2]: Do you mind putting these changes in its own commit, they are not really related to the whole PR. And you can add the argument for doing this to the commit message.

File: cpu/riscv_common/Makefile.features [hunk 3]
```diff
@@ -1,3 +1,5 @@
+CPU_ARCH := risc-v
```
Comment (fjmolinas) [inline cpu/riscv_common/Makefile.features:1]: Same for this one, can you add to a commit where you say "adding missing arch or something like that", or fix missing arch and you can ad comment on the `:=` change in that commit as well.

File: Makefile.include [hunk 4]
```diff
@@ -394,6 +394,9 @@ include $(BOARDDIR)/Makefile.include
 INCLUDES += -I$(RIOTCPU)/$(CPU)/include
 include $(RIOTCPU)/$(CPU)/Makefile.include
 
+# include architecture specific settings
+include $(RIOTMAKE)/arch/$(CPU_ARCH).inc.mk
```
Comment (fjmolinas) [inline Makefile.include:398]:
```suggestion
-include $(RIOTMAKE)/arch/$(CPU_ARCH).inc.mk
```

With this we can avoid the empty files I believe.

**Review by fjmolinas — COMMENTED** [review]
This one looks good, there are just two changes I would put in different commits. This is a nice cleanup and I don't expect any impact since thee include still happens after the `CPU` one so any board/cpu specific configuration can still override the generic one.

**Comment by maribu** [comment]
Done. And a rebase seems to also be needed due to a merge conflict.

**Review by fjmolinas — APPROVED** [review]
ACK, lets see what the ci says.

**Comment by fjmolinas** [comment]
Sorry @maribu I forgot to check the ci result and now this one needs a rebase :(

**Comment by maribu** [comment]
Strange, `git` was able to auto-merge locally during rebase.

**Review by fjmolinas — APPROVED** [review]
ACK

**Comment by benpicco** [comment]
looks like mips is missing

**Comment by maribu** [comment]
> looks like mips is missing

No, it worked fine prior to the XFA merge. The issue seems to be that the order of linker flags is slightly different with this PR for MIPS. All other arch but MIPS include `makefiles/arch/$(CPU_ARCH).inc.mk` at the end of `cpu/$(CPU)/Makefile.include`. In MIPS, it is included at the top of that file. Apparently, order didn't matte before, but with XFA it does now.

I will investigate this later. I believe there is some value in having the toolchain, CFLAGS and LDFLAGS setup as consistent as possible between all archs, as this eases refactoring and maintaining stuff.

**Comment by OlegHahm** [comment]
@maribu, ping :)

**Comment by maribu** [comment]
This is the type of PRs that horribly break since in a way that Murdock won't catch it, e.g. by triggering a build system bug that ends up half of the test cases not being build anymore. I agree that I should get back to this PR and push it forward, but maybe it is better to wait for the next development cycle before I break everything :)

**Comment by gschorcht** [comment]
@maribu Sorry for jumping into this discussion so late. To be honest, I would prefer to drop the changes for ESPs for now, as they conflict with PRs #17842, #17844, and the PR for ESP32S3 support that is hopefully coming in the next few days. In these PRs `xtensa` or `riscv-esp32` first define the CPU architecture and `xtensa-lx6`, `xtensa-lx7` and `riscv32` then define the CPU core used but not the architecture:
```
ifneq (,$(filter esp32,$(CPU_FAM)))
  CPU_ARCH = xtensa
  CPU_CORE = xtensa-lx6
else ifneq (,$(filter esp32c3,$(CPU_FAM)))
  CPU_ARCH = riscv_esp32
  CPU_CORE = riscv32
else ifneq (,$(filter esp32s3,$(CPU_FAM)))
  CPU_ARCH = xtensa
  CPU_CORE = xtensa-lx7
else
  $(error Unkwnown ESP32x SoC variant (family))
endif
```
Merging your changes now would cause a lot of conflicts when these very large PRs are rebased, and I would have to check everything again to make sure it still works. So I would prefer to clean it up later if it is still necessary.

**Comment by maribu** [comment]
OK, let's put this on hold until https://github.com/RIOT-OS/RIOT/pull/17842 and https://github.com/RIOT-OS/RIOT/pull/17844 are in.

**Comment by gschorcht** [comment]
> OK, let's put this on hold until #17842 and #17844 are in.

@maribu The changes of these PRs are merged now but there are a number of new PRs (#18410, #18411, ...) for ESP32-S3 support. Since the PR seems to diverge already, the question is whether we should get this PR merged before #18410, #18411. It will require of changes for ESPs.

**Comment by maribu** [comment]
I currently have no time to push this PR. Let's get ESP32-C3 in first. I can rebase and adapt this afterwards when I have time.
